Transaction integration tests

piotr-iohk opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->


# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [x] I want to add some transaction integration tests for api and cli 


# Comments

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Acknowledge any changes required to the Wiki
-->

paweljakubas commented:
This was only to make sure ocassional CI failure is not recurring

Commit:
more extraction to TestData.hs

Commit:
test transaction to mutliple addresses

piotr-iohk commented on lib/http-bridge/test/integration/Test/Integration/Scenario/Transactions.hs:
```diff
@@ -95,6 +100,126 @@ spec = do
+            ("POST", "v2/wallets/" <> widSrc <> "/transactions")
+            Default payload
+        verify r
+            [ expectResponseCode HTTP.status202
```
I don't know why this request is giving me `403`. Everything works fine with `curl`....

KtorZ replied:
you can't do multi-output transactions at the moment with those faucet wallets. That's something I wanted to enable yesterday but time flies. Since there's only a single UTxO on the faucet wallet at the moment, you can only have a single output recipient :) 
We need to send more than one output on the faucet wallet and have let's say, one million ADA, but not as 1x1M, but 10x100_000 !

piotr-iohk commented on lib/http-bridge/test/integration/Test/Integration/Scenario/Transactions.hs:
```diff
@@ -95,6 +100,126 @@ spec = do
+        --post transaction
+        r <- request @ApiTransaction ctx ("POST", postTx wa) Default payload
+        verify r
+            [ expectResponseCode HTTP.status202
```
Same here... `403` ¯\_(ツ)_/¯

KtorZ committed:
fix tx size estimation! Initial assumption about tx index being < 23 was not true .... :facepalm:

KtorZ committed:
generate 10x 100_000 Ada instead of 1x 1_000_000 for fixture wallets

KtorZ committed:
extend DSL with 'expectFieldBetween'

KtorZ committed:
have 'createWallet' from the integration DSL return an APIWallet (instead of an id)

KtorZ committed:
Adjust integration tests transaction scenarios to cover multi-outputs

Fees are actually not necessarily fixed. They depend on the ix of the input picked. Since we generate all fixture wallets
with an initial transaction with 1K outputs, ix can go from 0 to 999, and therefore be encoded on 1, 2 or 3 bytes. This
depends on which input are assigned to which wallet. As a consequence, it is easier (and more correct) to look at a range
of fees, in case inputs ix gets encoded on 1 or 3 bytes

Commit:
negative test for multi-output tx

KtorZ committed:
factor-out some logic related to creating fixture wallets

KtorZ committed:
add more negative transactions scenarios

KtorZ committed:
add CLI scenario for testing sending a transaction via CLI

KtorZ reviewed (approved):
:ok_hand:
